Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs

```diff
@@ -81,50 +106,99 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
         }
 
-        internal async Task<Page> GetNextPage(long lastProcessedCheckpoint, string subscriptionId)
+        internal void Unsubscribe(Subscription subscription)
         {
-            if (isDisposed)
+            lock (subscriptionLock)
+            {
+                subscriptions.Remove(subscription);
+            }
+        }
+
+        internal async Task<Page> GetNextPage(long precedingCheckpoint, string subscriptionId,
+            CancellationToken cancellationToken)
+        {
+            if (disposingCancellationTokenSource.IsCancellationRequested)
             {
                 throw new ObjectDisposedException(typeof(PollingEventStoreAdapter).FullName);
             }
 
-            Page pageFromCache = TryGetNextPageFromCache(lastProcessedCheckpoint, subscriptionId);
+            Page pageFromCache = TryGetNextPageFromCache(precedingCheckpoint, subscriptionId);
             if (pageFromCache.Transactions.Count > 0)
             {
                 return pageFromCache;
             }
 
+            LoadRequest pendingRequest =
+                pendingRequests.ToArray().SingleOrDefault(x => x.PrecedingCheckpoint == precedingCheckpoint);
+
+            if (pendingRequest != null)
+            {
+                logDebug(() =>
+                    $"Pending Request Found (subscription: {subscriptionId}, originalSubscription: " +
+                    $"{pendingRequest.SubscriptionId}, preceding checkpoint: {precedingCheckpoint}");
+
+                return await pendingRequest.Page;
+            }
 
-            Page loadedPage = await LoadNextPage(lastProcessedCheckpoint, subscriptionId).ConfigureAwait(false);
- 
-            if (loadedPage.Transactions.Count == maxPageSize)
+            var completedPage = await RequestPageload(subscriptionId, precedingCheckpoint, cancellationToken);
+            if (ShouldPreloadNextPage(completedPage))
             {
-                StartPreloadingNextPage(loadedPage.LastCheckpoint, subscriptionId);
+                RequestPageload(subscriptionId, completedPage.PrecedingCheckpoint, cancellationToken);
             }
 
-            return loadedPage;
+            return completedPage;
+        }
+
+        private Task<Page> RequestPageload(string subscriptionId, long precedingCheckpoint, CancellationToken cancellationToken)
+        {
+            var loadRequest = new LoadRequest
+            {
+                SubscriptionId = subscriptionId,
+                PrecedingCheckpoint = precedingCheckpoint,
+                CancellationToken = cancellationToken
+            };
+            
+            pendingRequests.Enqueue(loadRequest);
+            requestQueued.TrySetResult(true);
+
+            return loadRequest.Page;
+        }
+
+        private bool ShouldPreloadNextPage(Page completedPage)
+        {
+            // If there's no cache configured, preloading will only benefit if the subscriber requests the next page while it is still in the queue.

```

Comment:
I can't relate this comment to the code below.

Reply:
Tried to rephrase it